storage/src/main/java/zipkin2/storage/kafka/KafkaSpanConsumer.java

```diff
@@ -49,12 +53,20 @@
   @Override
   public Call<Void> accept(List<Span> spans) {
     if (spans.isEmpty()) return Call.create(null);
-    List<Call<Void>> calls = new ArrayList<>();
     // Collect traceId:spans
+    Map<String, List<Span>> spansByTrace = new HashMap<>();
     for (Span span : spans) {
       String key = span.traceId();
-      byte[] value = SpanBytesEncoder.PROTO3.encodeList(Collections.singletonList(span));
-      calls.add(KafkaProducerCall.create(producer, spansTopicName, key, value));
+      List<Span> current = spansByTrace.get(key);
+      if (current == null) current = new ArrayList<>();
+      if (!current.contains(span)) current.add(span);
+      spansByTrace.put(key, current);
+    }
+    // Serialize spans by trace Id
+    List<Call<Void>> calls = new ArrayList<>();
```

Comment:
Not sure it helps you but was have a function GroupByTraceId. one thing is that spans could be mixed length trace IDs.. pain I know.. This may lead you to want to partition by the 64 bits

Reply:
applied.